Separate dense and decomposed footprint API

lagru opened the pull request:
## Description

Closes #7736; see that issue for more context. I'll expand this description as this PR shapes up.

- Exposes `_cross_decomposition` as a new public `footprint_cross_decompose`.

## Checklist

<!-- Before pull requests can be merged, they should provide: -->

- A descriptive but concise pull request title
- [Docstrings for all functions](https://github.com/numpy/numpydoc/blob/main/doc/example.py)
- [Unit tests](https://scikit-image.org/docs/dev/development/contribute.html#testing)
- A gallery example in `./doc/examples` for new features
- [Contribution guide](https://scikit-image.org/docs/dev/development/contribute.html) is followed

## Release note

For maintainers and optionally contributors, please refer to the [instructions](https://scikit-image.org/docs/stable/development/contribute.html#documenting-changes) on how to document this PR for the release notes.

```release-note
Add the new function `skimage.morphology.footprint_cross_decompose`, which can separate
symmetric convex 2-dimensional footprints into an equivalent series of smaller cross-shaped
footprints. {label="New feature"}
```

```release-note
In `skimage.morphology.footprint_rectangle`, deprecate the argument `decomposition`;
instead use the new function `skimage.morphology.footprint_decomposed_rectangle` to 
create a series of footprints, that combine to an equivalent rectangular footprint. {label="API"}
```

lagru committed:
Make footprint cross decomposition a public function

Since this function is now exposed to all kinds of footprints, I tried
to add sanity checks that didn't impact the performance to much and
tried to come up with a sensible behavior for edge cases.

It also removes the `dtype` parameter. Instead it just uses the dtype
of `footprint`, allowing users to control the outputs dtype that way.
I couldn't really think of use cases where the parameter would be
more useful.

lagru commented on skimage/morphology/footprints.py:
```diff
@@ -612,41 +612,127 @@ def _cross(r0, r1, dtype=np.uint8):
+        A symmetric convex footprint as a 2-D array of 1's and 0's.
+        N and M must be of odd length.
+
+        .. warning::
+            For performance reasons, it isn't checked whether `footprint` is
+            symmetric or convex or contains only 0 and 1. So, for unsupported
+            footprints, this function may return an incorrect decomposition!
```
I'd appreciate early feedback here. I'm not aware of a fast way to check if large footprints are convex and symmetric. So instead I tried to add performance-friendly checks where I could and added this warning.

What do you think?

matthew-brett replied:
Is there a good way of leaving this choice of slow-accurate vs fast-dangerous to the user?

lagru replied:
I thought about using a parameter like `check_convex_symmetry : bool` that defaults to true. For large footprints, users could disable that one. 

Or we could only check smaller footprints where the performance isn't impacted too much. And if users want explicit checking behavior the can use `check_convex_symmetry`...?

lagru replied:
We could also just keep this and these kind of functions private and only duplicate the existing functions with the `decomposition=` parameter. It's a trade-off between an API (a) that gives users more low-level building blocks that have to be used with care and an API (b) with fewer safer building blocks that might not support all use cases of (a).

lagru replied:
@stefanv suggested that may we could simply check that the decomposed footprint composes again to the dense version at the end of the function. This shouldn't impact performance too much in the big picture and if it does. I'll come up with a proposal.

lagru committed:
Add max_error argument to footprint_cross_decompose

This should give users control on how correct the approximation is
supposed to be and also catches cases where the footprint isn't convex.

lagru committed:
Check symmetry in footprint_cross_decompose

lagru committed:
Expand docstring of footprint_cross_decompose

lagru commented on skimage/morphology/footprints.py:
```diff
@@ -612,41 +612,166 @@ def _cross(r0, r1, dtype=np.uint8):
+            f"each dimension, got `{footprint.shape=}`"
+        )
+        raise ValueError(msg)
+    for dim in range(footprint.ndim):
+        mirrored = np.flip(footprint, axis=dim)
+        if np.any(footprint != mirrored):
+            msg = f"`footprint` isn't symmetric in dimension {dim}"
+            raise ValueError(msg)
```
With the addition of the `max_error` argument, this check isn't strictly necessary since a non-convex footprint should usually exceed the allowed error rate. However, we may want to keep this because it provides a clearer warning?

lagru committed:
Fix typo

lagru committed:
Don't cross decomposition error in disk & ellipse

lagru committed:
Expand docstring of _decompose_size

lagru committed:
Create new `footprint_decomposed_rectangle` function

and deprecate `decomposition` argument in `footprint_rectangle`. It's
a bit unfortunate that we need to do a deprecation warning because
it was already released in 0.25.

lagru committed:
Remove accidental committed draft of `footprint_compose`

lagru committed:
Add `dtype` argument to `footprint_from_sequence`

lagru committed:
Revert "Add `dtype` argument to `footprint_from_sequence`"

This reverts commit 66c399818a42d51eb86dd67c670af75abd0d6673.

lagru commented on skimage/morphology/footprints.py:
Note the function `footprint_from_sequence(footprints)` which is a bit of an outlier in terms of dtype behavior. Unlike all other footprint functions, it doesn't allow specifying the dtype and it returns a `bool` array by default; all others default to `np.uint8`.

If we want to change this default behavior, we'd have to do so in skimage2. Unless we rename the function to something like `footprint_compose` (which I prefer). 

Thoughts?

lagru committed:
Fix plot_footprint_decompositions gallery example

lagru committed:
Merge branch 'main' into separate_footprint_decomposition

stefanv commented:
Thinking again about the naming here, we may want to go for `footprint_rectangle`, `footprint_rectangle_decomposed` (which I think was how you originally had it). Sorry that this got out of sync with main before being reviewed.

matthew-brett commented:
We have changed the base branch for this PR from the `main` branch to the `pre-megamove-integration` branch.

## Summary - what this means for your PR

You can continue to work on your PR, now based on `pre-megamove-integration`.  You should update your branch to the current state of `pre-megamove-integration` (this is what the Github "Update branch" button does).   Please **do not merge the scikit-image `main` branch into your PR branch**.

Once we've merged your PR into the `pre-megamove-integration` branch, we'll update your changes for Megamove, and merge into the `main` branch.

In addition - our automated script detected that your branch had merge conflicts with the pre-Megamove branch. Please make sure to merge in the `pre-megamove-integration` branch, and resolve the merge conflicts in the way that you want.

Read on for the Megamove rationale, and more detailed instructions.

<details>

## Rationale

We've changed the base branch to make it easier and smoother to continue to work on your PR, while we work through a rather large change in the directory structure of the repository - a change that we have called *Megamove*.

See [the Megamove explanation issue](https://github.com/scikit-image/scikit-image/issues/8232) for more detail.

Changing the base branch means you do not see merge conflicts related to the *Megamove* change, and can continue to work with the repository in its original state, without worrying about porting those changes across to the new post-Megamove structure.

As you'll see in the linked issue, you can continue to work on your PR as before, and you can continue to use the Github GUI to "Update branch" - this updates from the new `pre-megamove-integration` branch.   However, there is one caveat: please, do not try and merge the `scikit-image` `main` branch into this one, but instead, keep track of the `pre-megamove-integration` branch.  The issue explains more.

To emphasize again, **do not try and merge the scikit-image `main` branch**.  You will get Megamove related merge conflicts that are rather tricky to resolve correctly.

When we've approved your PR, we'll merge into the `pre-megamove-integration` branch, and then process it so we can merge it into the `main` branch.   We'll let you know when we've merged your PR changes into `main`.

</details>

Any questions, please reply here.
